### ISSUE TITLE
[tune] lazily generate configurations for grid search

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#10138]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

This currently breaks on most setups.

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):
```
from ray import tune


def objective(step, alpha, beta):
    return (0.1 + alpha * step / 100)**(-1) + beta * 0.1


def training_function(config):
    # Hyperparameters
    alpha, beta = config["alpha"], config["beta"]
    for step in range(10):
        # Iterative training function - can be any arbitrary training procedure.
        intermediate_score = objective(step, alpha, beta)
        # Feed the score back back to Tune.
        tune.report(mean_loss=intermediate_score)


analysis = tune.run(
    training_function,
    config={
        "alpha": tune.grid_search([0.001, 0.01, 0.1]),
        "beta": tune.choice(list(range(10000)))
    },
    num_samples=1000000)
```
https://stackoverflow.com/questions/63395596/ray-tune-random-search-indefinitely-many-samples

**Comment by krfricke** [comment]
So the problem here lies in two places. For one, the code above does request scheduling of 3 million runs in parallel. There is no easy way to limit concurrency here, see also #10256.

The other thing is that the basic variant generator generates all configurations at once, and creates trial objects for all of these. This could be fixed by batching the creation, like in this proof-of-concept commit:

https://github.com/krfricke/ray/commit/74ef52d8cc7e5b99e567cfbba11699547a95e842

This would however implicitly limit concurrency to `batch_size`. Thus it would make sense to combine this with #10256.

**Comment by richardliaw** [comment]
This is pretty interesting - there's actually 3 different rate-limiting steps to be aware of.

1. Sampling concurrency (currently set to infinite, bounded by ConcurrencyLimiter): This is important because it affects the actual optimization.
2. Machine level concurrency (currently set to total_resources / resources_per_trial): This is important because it prevents resource exhaustion/overloading of trials
3. Autoscaling queue depth (defaults to 1): This is important (later on) because it will guide autoscaling decisions.

Another important consideration is "out of band" trials. That is, it is possible for sampling concurrency = 1, but the actual number of concurrent evaluations = N > 1. (See Repeater, or the Tune Client API).

Another important consideration is autoscaling. We want to support the case where we have 0 GPUs in our cluster (CPU-only head node, with autoscaling workers). This will require a trial to *start* even when the machine level concurrency = 0 (effectively).

Therefore, my recommendation is the following:

* For search algs/grid search:
  * Only generate trials when asked, 1 by 1.
  * Return None if sampling concurrency is met.
* For trial runner:
  * Do not greedily gather/collect trials - only query up to "queue depth + machine level concurrency" trials at once.

RE: `Machine level concurrency: int` - having a parameter for this only makes sense for fixed cluster sizes, but also not sure how many people actually need to set this. It most likely will be overriden by sampling concurrency in all cases.

**Comment by krfricke** [comment]
Closed via #10802
